### PR TITLE
Sample pooling

### PR DESCRIPTION
**Pull request by josenavas**
Adds the GUI for pooling samples.

pooling

### CODE REVIEW

File: knimin/templates/pm_targeted_pool.html [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+   * @param {Element} input The inpute element that has been updated
+   *
+   **/
+  function runPoolVolumeUpdated(input) {
+    var currentEntry = $(input).parents('.entry:first');
+    var runTotalPoolVolume = parseFloat($('#run-pool-volume-input').val());
+    var runPoolVolume = parseFloat($(input).val());
+    // Magic number 10 -> pool-input is in micrograms, so we need
+    // to multiply 1000 to get nanograms (pool plate volume), since
+    // we have to divide by 100 due to the percentage value: 1000/100 = 10s
```
Comment (antgonza) [inline knimin/templates/pm_targeted_pool.html]: Should these values be global variables?

Comment (josenavas) [inline knimin/templates/pm_targeted_pool.html]: I don't think it is worth adding them as global variables.

Comment (antgonza) [inline knimin/templates/pm_targeted_pool.html]: k, thanks. 👍  once tests pass.

**Comment by josenavas** [comment]
@qiyun can you take a look to this one?

File: knimin/handlers/pm_pool_handlers.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from knimin.handlers.base import BaseHandler
+from knimin.handlers.access_decorators import set_access
+from knimin import db
+
+
+@set_access(['Admin'])
+class PMPoolPlatesHandler(BaseHandler):
+    @authenticated
+    def get(self):
+        arg_plates = map(int, self.get_arguments('plate'))
```
Comment (qiyunzhu) [inline knimin/handlers/pm_pool_handlers.py]: maybe `plate_args` is a better name for this variable?

File: knimin/lib/data_access.py [hunk 3]
```diff
@@ -3654,6 +3654,150 @@ def delete_targeted_plate(self, plate_id):
...
+            The total volume of the pool
+
+        Returns
+        -------
+        int
+            The run pool id
+
+        Raises
+        ------
+        ValueError
```
Comment (qiyunzhu) [inline knimin/lib/data_access.py]: Where is this error raised?

File: knimin/lib/tests/test_data_access.py [hunk 4]
```diff
@@ -1294,6 +1294,164 @@ def test_delete_targeted_plate(self):
...
+
+        # Create the layout
+        layout = []
+        row = []
+        for i in range(pt['rows']):
+            for j in range(pt['cols']):
+                row.append({'sample_id': None, 'name': None, 'notes': None})
+            layout.append(row)
+            row = []
+        layout[0][0]['sample_id'] = samples[0]
```
Comment (qiyunzhu) [inline knimin/lib/tests/test_data_access.py]: Trivial. This chuck of code can be simpler.

File: knimin/lib/tests/test_data_access.py [hunk 5]
```diff
@@ -1294,6 +1294,164 @@ def test_delete_targeted_plate(self):
...
+        self._clean_up_funcs.insert(
+            0, partial(db.delete_sample_plate, plate_id_2))
+
+        # Plate some samples
+        # Add samples to the study
+        samples = ['9999.Sample_1', '9999.Sample_2', '9999.Sample_3',
+                   '9999.Sample_3']
+        db.set_study_samples(9999, samples)
+
+        # Create the layout
```
Comment (qiyunzhu) [inline knimin/lib/tests/test_data_access.py]: This piece of code is used multiple times. Maybe create a private function for it?

File: knimin/templates/pm_targeted_pool.html [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+                         '</div>' +
+                         '<div class="col-xs-1 form-group">' +
+                           '<button class="btn btn-info btn-remove form-control" type="button"><span class="glyphicon glyphicon-minus"></span></button>' +
+                         '</div>' +
+                       '</div>');
+
+      addPlateOptions(newEntry.find('.plate-select:first'));
+      newEntry.appendTo(controlForm);
+      initGUI();
+    }).on('click', '.btn-remove', function(e) {
```
Comment (qiyunzhu) [inline knimin/templates/pm_targeted_pool.html]: Oh I didn't know that one can write two event listeners in one line :)

File: knimin/templates/pm_targeted_pool.html [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+      e.preventDefault();
+      initGUI();
+      return false;
+    });
+  });
+</script>
+{% end %}
+{% block content %}
+<h3>Prepare targeted pool</h3>
+<div class="container">
```
Comment (qiyunzhu) [inline knimin/templates/pm_targeted_pool.html]: It looks to me that some HTML codes are duplicated? e.g., `<div class="entry row">`

File: knimin/handlers/pm_pool_handlers.py (hunk 2 above)
Comment (josenavas) [inline knimin/handlers/pm_pool_handlers.py]: Done

File: knimin/lib/data_access.py (hunk 3 above)
Comment (josenavas) [inline knimin/lib/data_access.py]: Forgot to actually add the check - added!

File: knimin/lib/tests/test_data_access.py (hunk 4 above)
Comment (josenavas) [inline knimin/lib/tests/test_data_access.py]: I don't think is critical, but please provide the alternative.

File: knimin/lib/tests/test_data_access.py (hunk 5 above)
Comment (josenavas) [inline knimin/lib/tests/test_data_access.py]: Done

File: knimin/templates/pm_targeted_pool.html (hunk 6 above)
Comment (josenavas) [inline knimin/templates/pm_targeted_pool.html]: 😄

File: knimin/templates/pm_targeted_pool.html (hunk 7 above)
Comment (josenavas) [inline knimin/templates/pm_targeted_pool.html]: yes, those are defining the multiple rows in the interface, it is using [bootstrap's grid system](https://v4-alpha.getbootstrap.com/layout/grid/) (which is awesome)

File: knimin/lib/tests/test_data_access.py (hunk 4 above)
Comment (qiyunzhu) [inline knimin/lib/tests/test_data_access.py]: Not critical at all. Alternative:
```
for i in range(4):
    layout[0][i]['sample_id'] = samples[i]
for i in reversed(range(2)):
    db.write_sample_plate_layout(plate_id, layout[:-i])
```

**Comment by qiyunzhu** [comment]
Looks good now! @josenavas

File: knimin/lib/tests/test_data_access.py (hunk 4 above)
Comment (antgonza) [inline knimin/lib/tests/test_data_access.py]: As it's not critical, to speed up things, and as the rest has been approved: I'm gonna merge.

**Pull request merged**